### PR TITLE
Add spreadable spread hooks

### PR DESCRIPTION
**Pull request by variousauthors**
This PR adds the ability to control how grass and mycelium spread. It adds an event `SpreadableSpreadEvent` which has a pre and post phase, as well as hooks that are called by the `SpreadableSnowyDirtBlock`'s `tick` method.

I think this functionality is useful and fun! In my test mod file I have given two examples: one of them changes the way mycelium spreads (it spreads grey wool instead of mycelium). The other example adds a new dirt/grass pair (special dirt and special grass) which behave like dirt and grass but do not interact with dirt and grass.

My concrete use case for wanting to add this hook to forge was to create a block that functions like farmland but looks and spreads like grass. I like having idyllic gardens, and farmland looks to agricultural. I want to have wild pumpkin stalks in my idyllic gardens. When I discovered that I couldn't do this, I went on the forge discord and it was suggested that I make a PR.

One could use this hook to make cool blocks that behave like grass, such as something like this [grassy stone][0] from the forums.

Here is an example of someone else asking for this functionality [on the forum][1]. 

Here is another PR providing basically the same functionality, made back in October: https://github.com/MinecraftForge/MinecraftForge/pull/6214

---

Now, with regards to that last point. I probably should have searched the open PRs before starting work on this. The PR above is broader in scope than this one: it treats fire, mushrooms, sea pickles, and such as "spreadable" in a general sense. When investigating the code I see that, indeed, they do all work pretty much the same. My change is much smaller, and could easily be expanded to include the other spreadable blocks later.

So! I'm going to open this PR anyway, but I'll understand if it is "closed as a duplicate" or something.

[0]: https://www.minecraftforge.net/forum/topic/28095-1710-making-custom-quotgrassquot-block-with-all-of-its-functions/
[1]: https://www.minecraftforge.net/forum/topic/63956-112-make-custom-grass-block-use-the-colour-map/

### CODE REVIEW

**Comment by variousauthors** [comment]
Here is a link to the tiny mod that will use these hooks: 

https://github.com/variousauthors/FertileSoil

File: src/main/java/net/minecraftforge/common/extensions/IForgeSpreadable.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.extensions;
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/common/extensions/IForgeSpreadable.java:20]: I think this belongs to the net.minecraftforge.common, as it's a specific feature as IShearable and IPlantable

**Comment by variousauthors** [comment]
I published the tiny mod that will one day use these hooks (for now my custom dirt searches for nearby grass and spreads it manually).

https://www.curseforge.com/minecraft/mc-mods/fertile-soil

File: src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java [hunk 2]
```diff
@@ -523,6 +523,25 @@ default boolean addDestroyEffects(BlockState state, World world, BlockPos pos, P
...
+    /**
+     * Determines if this block can support the passed in spreadable, allowing it to spread to this block.
+     * 
+     * The only spreadables in vanilla minecraft are Grass and Mycellium, and they can both only
+     * be spread to Dirt.
+     * 
+     * @param spreadable
+     * @return True to allow the spreadable to spread to this block.
+     */
+    default boolean canSustainSpreadableOfType(Block spreadable)
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:535]: Is it possible to change this to a block state for more context?

**Review by ichttt — CHANGES_REQUESTED** [review]

**Pull request closed without merging**